### DISCUSSION TITLE
Paragraph level / block level indexer for sentence level backlink

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#7537] [Ideas]
I'm thinking about making a back-line-link-indexer, that works like backlink-indexer, but can do RoamResearch/Obsidian/Logseq style block level link [[Index^aw35awf]] Title^hash-of-line.

And it may also work for line level transclusion.

This feature is useless when user not using outline a lot, but with WYSIWYG editor, typing a list is much smoother, so this might be a useful feature.

P.S. I can't search out an issue about this, using keywords in the title, but I remember I have seen one years ago.

PR: https://github.com/Jermolene/TiddlyWiki5/pull/7744

**Comment by Jermolene** [comment]
Hi @linonetwo you'd like to make individual lines within a tiddler be addressable via links of the form `{title}^{hash of line}`, is that correct?

How is the hash computed? Does the hash change if the text of the line changes, or are hashes remembered somehow?

I understand this need, and it crops up in various forms fairly often. The TiddlyWiki way of addressing this need would be to put each block into its own tiddler. 

I've always felt that attempts to add support for addressable units within tiddlers would gradually end up re-inventing the tiddler, ending up with duplication in the design.

**Reply by linonetwo** [reply]
I was describing obscurely in the discussion because I think someone will link to their previous issues later, so this is a discussion not an issue.

But https://github.com/Jermolene/TiddlyWiki5/discussions/7537#discussioncomment-6165584

**Comment by linonetwo** [comment]
But if you ask me to give a solution immediately: Let's just use **partially Obsidian compatible** way, which is also text-based, and proven to work stably:

## Use hash

1. Hash won't change, it is written by hand, and it is a ` \^\S+$` string after line: `text ^cb9d485` or `text ^1`, so it can be human readable without space, we can add a new parse rule for this.
2. And omit this hash when creating widget for rendering, so it's invisible in view mode. But this widget will create an anchor to jump to.
3. link looks like `[[Test1#^762286]]`, use `#` for separator, next part is title name or `^` + hash above. And also, we will make `#^762286` part invisible in Link widget.
4. `[[Test1#111#asdf]]` will goto a level 2 title `asdf`

What is different to Obsidian, is our `|` is before the tiddler name, Obsidian is after: `[[Test1#^acb9d|Shown name]]`, so this is only "partially compatible", we have to blame them to didn't do it the compatible way to wikitext :P

## Error handling

1. when a user modifies the hash or title, or the hash/title in the link is wrong: just jump to the tiddler title directly, as the old way.
2. when there is no space between the text and the hash, or hash contains space: just ignore this syntax, show as normal text
3. multiple same hash in a file: jump to the last occurrence. Also, this hash is "namespaced" by tiddler title, so don't need to worry about its duplication across files.

## Virtual tiddler

> I've always felt that attempts to add support for addressable units within tiddlers would gradually end up re-inventing the tiddler, ending up with duplication in the design.

Maybe we can just call these "line with ^hash" as "Virtual tiddler" or "Derived tiddler" or "Virtual Fragment tiddler", and they have fields:

- "virtual-parent" point to the real tiddler
- "title" field is the link to it `Test1#^acb9d`
- "text" field is the line text.

We construct them during startup, just like how we build indexer at startup, which can't be defered #7277 

In this way, most of things will work out-of-box...Just need to prevent them get sync to twcloud / backup to git / store to fs, by checking the "virtual-parent" field, like we check the `canonical_uri`.

## Update all other things

The Backlink indexer needs an update, luckily that it is constructed at runtime, and the index is a private member (is it?), so we don't need to worry about the back compatibility.

https://github.com/Jermolene/TiddlyWiki5/blob/f90eb386ae7cd7da0f18b8d53fb0e454d9486d75/core/modules/indexers/backlinks-index.js#L73

Also need to track block level link. So Need two indexers.

## Other utils 

- wiki.extractLinks

...TBD

**Reply by linonetwo** [reply]
@Jermolene I'm building this, so I can implement a foldable-list, that stores the folding state in a field, with block id as key.

**Reply by Jermolene** [reply]
Hi @linonetwo that's an interesting direction to explore, and I'll look forward to seeing the results of your experiments.

Your proposal contains the proposal of a wikitext shortcut syntax for creating link anchors which I like, and think would be viable. 

I am not so keen on the virtual tiddler idea for the reasons given above.

**Reply by linonetwo** [reply]
@Jermolene Thanks for confirmation.

About how to show a block in the "Backlinks", I rethink about it, while writing https://github.com/Jermolene/TiddlyWiki5/pull/6081 , I also agree that "virtual tiddler" is not necessary.
We can just get the block content from AST on the fly, no need to cache it, nor need an indexer for it, because this will only happened on `O( {{{ [backlink[]] }}} ) = O(1)` tiddlers, it is very fast.
